src/algebra/gcd_monoid.lean

```diff
@@ -600,3 +614,149 @@ lemma gcd_eq_of_dvd_sub_left {a b c : α} (h : a ∣ b - c) : gcd b a = gcd c a
 by rw [gcd_comm _ a, gcd_comm _ a, gcd_eq_of_dvd_sub_right h]
 
 end integral_domain
+
+section constructors
+variables [comm_cancel_monoid_with_zero α] [nontrivial α] [normalization_monoid α]
+
+/-- Define `gcd_monoid` on a structure just from the `gcd` and its properties. -/
+noncomputable def gcd_monoid_of_gcd [decidable_eq α] (gcd : α → α → α)
+  (gcd_dvd_left   : ∀a b, gcd a b ∣ a)
+  (gcd_dvd_right  : ∀a b, gcd a b ∣ b)
+  (dvd_gcd        : ∀{a b c}, a ∣ c → a ∣ b → a ∣ gcd c b)
+  (normalize_gcd  : ∀a b, normalize (gcd a b) = gcd a b) :
+  gcd_monoid α :=
+{ gcd := gcd,
+  gcd_dvd_left := gcd_dvd_left,
+  gcd_dvd_right := gcd_dvd_right,
+  dvd_gcd := λ a b c, dvd_gcd,
+  normalize_gcd := normalize_gcd,
+  lcm := λ a b, if a = 0 then 0 else classical.some (dvd_normalize_iff.2
+          (dvd.trans (gcd_dvd_left a b) (dvd.intro b rfl))),
+  gcd_mul_lcm := λ a b, by {
+    by_cases a0 : a = 0,
+    { rw [if_pos a0, mul_zero, a0, zero_mul, normalize_zero] },
+    { rw if_neg a0,
+      exact (classical.some_spec (dvd_normalize_iff.2
+                  (dvd.trans (gcd_dvd_left a b) (dvd.intro b rfl)))).symm } },
+  lcm_zero_left := λ a, if_pos rfl,
+  lcm_zero_right := λ a, by {
+    by_cases a0 : a = 0, { apply if_pos a0 },
+    rw if_neg a0,
+    rw ← normalize_eq_zero at a0,
+    have h := (classical.some_spec (dvd_normalize_iff.2
+                  (dvd.trans (gcd_dvd_left a 0) (dvd.intro 0 rfl)))).symm,
+    have gcd0 : gcd a 0 = normalize a,
+    { rw [← normalize_gcd],
+      apply normalize_eq_normalize (gcd_dvd_left _ _) (dvd_gcd (dvd_refl a) (dvd_zero a)) },
+    rw ← gcd0 at a0,
+    apply or_iff_not_imp_left.1 (mul_eq_zero.1 _) a0,
+    rw [h, mul_zero, normalize_zero] },
+  .. (infer_instance : normalization_monoid α) }
+
+/-- Define `gcd_monoid` on a structure just from the `lcm` and its properties. -/
+noncomputable def gcd_monoid_of_lcm [decidable_eq α] (lcm : α → α → α)
+  (dvd_lcm_left   : ∀a b, a ∣ lcm a b)
+  (dvd_lcm_right  : ∀a b, b ∣ lcm a b)
+  (lcm_dvd        : ∀{a b c}, c ∣ a → b ∣ a → lcm c b ∣ a)
+  (normalize_lcm  : ∀a b, normalize (lcm a b) = lcm a b) :
+  gcd_monoid α :=
+{ lcm := lcm,
+  gcd := λ a b, if a = 0 then normalize b else (if b = 0 then normalize a else
+    classical.some (dvd_normalize_iff.2 (lcm_dvd (dvd.intro b rfl) (dvd.intro_left a rfl)))),
```

Comment:
Since that expression appears a lot:
```suggestion
let exists_gcd := λ a b, dvd_normalize_iff.2 (lcm_dvd (dvd.intro b rfl) (dvd.intro_left a rfl)) in
{ lcm := lcm,
  gcd := λ a b, if a = 0 then normalize b else (if b = 0 then normalize a else
    classical.some (exists_gcd a b)),
```